code/modules/projectiles/guns/magnetic/bore.dm

```diff
@@ -77,91 +74,169 @@
+
+	icon_state = "bore"
+	item_state = "bore"
+	wielded_item_state = "bore-wielded"
+	one_handed_penalty = 5
+
+	projectile_type = /obj/item/projectile/bullet/magnetic/bore
+
+	gun_unreliable = 0
+	power_cost = 100
+	ammo_material = MAT_PHORON
+
+	action_button_name = "Toggle internal generator"
+
+	var/generator_state = GEN_OFF
+	var/datum/looping_sound/small_motor/soundloop
+	var/time_started //to keep the soundloop from being "stopped" too soon and playing indefinitely
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/Initialize()
+	. = ..()
+	soundloop = new(list(src), 0)
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/Destroy()
+	QDEL_NULL(soundloop)
 	. = ..()
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/ui_action_click()
+	toggle_generator(usr)
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/process()
+	if(generator_state && !mat_storage)
+		audible_message(SPAN_NOTICE("\The [src] goes quiet."),SPAN_NOTICE("A motor noise cuts out."))
+		soundloop.stop()
+		generator_state = GEN_OFF
+
+	else if(generator_state)
+		if(generator_state == GEN_IDLE && (cell?.percent() < 80 || (!cell && capacitor && capacitor.charge/capacitor.max_charge < 0.8)))
+			generator_state = GEN_ACTIVE
+		else if(generator_state == GEN_ACTIVE && (!cell || cell.fully_charged()) && (!capacitor || capacitor.charge == capacitor.max_charge))
+			generator_state = GEN_IDLE
+		soundloop.speed = generator_state
+		generator_generate()
+
+	if(capacitor)
+		if(cell)
+			if(capacitor.charge < capacitor.max_charge && cell.checked_use(power_per_tick))
+				capacitor.charge(power_per_tick)
+		else if(!generator_state)
+			capacitor.use(capacitor.charge * 0.05)
+
+	update_state()
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/proc/generator_generate()
+	var/fuel_used = generator_state == GEN_IDLE ? 5 : 25
+	var/power_made = fuel_used * 800 * CELLRATE //20kW when active, same power as a pacman on setting one, but less efficient because compact and portable
+	if(cell)
+		cell.give(power_made)
+	else if(capacitor)
+		capacitor.charge(power_made)
+	mat_storage = max(mat_storage - fuel_used, 0)
+	var/turf/T = get_turf(src)
+	if(T)
+		T.assume_gas("carbon_dioxide", fuel_used * 0.01, T0C+200)
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/proc/toggle_generator(mob/living/user)
+	if(!generator_state && !mat_storage)
+		to_chat(user, SPAN_NOTICE("\The [src] has no fuel!"))
+		return
+
+	else if(!generator_state)
+		generator_state = GEN_STARTING
+		var/pull = (!cell || cell.charge < 100) ? rand(1,4) : 0
+		while(pull)
+			playsound(src, 'sound/items/small_motor/motor_pull_attempt.ogg', 100)
+			if(!do_after(user, 2 SECONDS, src))
+				generator_state = GEN_OFF
+				return
+			pull--
+		soundloop.start()
+		cell?.use(100)
+		audible_message(SPAN_NOTICE("\The [src] starts chugging."),SPAN_NOTICE("A motor noise starts up."))
+		generator_state = GEN_IDLE
+
+	else if(generator_state > GEN_OFF && time_started + 3 SECONDS < world.time)
```

Comment:
You don't appear to set `time_started` anywhere. And this appears to just stop the generator 3 seconds after it's started?

Reply:
This prevents you from mashing the toggle hud icon to turn the generator off too quickly. If a soundloop is stop()ed while it's still playing the start sound, it doesn't stop and will keep going and I don't want to try and get into the soundloop code to try and change that right now.